File: 单调栈/最短无序连续子数组.go
package 单调栈

func findUnsortedSubarray(nums []int) int {
	stack := []int{0} //维护一个大顶栈，栈中存的是下标
	start := len(nums)
	end := 0
	for i := 1; i < len(nums); i++ {
		for len(stack) != 0 && nums[i] <= nums[stack[len(stack)-1]] { //为了维护大顶栈，nums[i] 必须要 换掉栈顶的元素
			start = min(start, stack[len(stack)-1])
			end = i
			stack = stack[:len(stack)-1]
		}
		stack = append(stack, i)
	}
	res := end - start + 1
	if res < 0 {
		return 0
	}
	return res
}

File: 贪心/最短无序连续子数组.java
package 贪心;

//部分排序
public class 最短无序连续子数组 {
    public int findUnsortedSubarray(int[] nums) {
        int n = nums.length;
        int left = -1, right = -2;
        var max = nums[0];
        var min = nums[n - 1];
        //从左到右的遍历记录最大值，如果当前值比最大值小，那么他就是右边界
        for (int i = 1; i < n; i++) {
            if (nums[i] >= max) max = nums[i];
            else right = i;
        }
        for (int i = n - 1; i >= 0; i--) {
            if (nums[i] <= min) min = nums[i];
            else left = i;
        }
        return right - left  +1;
    }
}
